app/models/katello/purpose_status.rb

```diff
@@ -1,8 +1,8 @@
 module Katello
   class PurposeStatus < HostStatus::Status
```

Comment:
Because of the changes in this file, I realize that for users in non-sca mode we are going to be telling them that their Matched hosts are Unknown and vice-versa, so we need a migration in order to correct the data. It's minor and should look something like below. Check out the rails docs on creating a migration if you've never done it before!

```
purpose_types = Katello::HostStatusManager::PURPOSE_STATUS.map(&:to_s)

# load both sets of host statuses and *then* update them to make sure we update the correct statuses
unknown_statuses = ::HostStatus::Status.where(type: purpose_types, status: 2)
matched_statuses = ::HostStatus::Status.where(type: purpose_types, status: 0)

unknown_statuses.update_all(status: Katello::PurposeStatus::UNKNOWN) # 2 => 0
matched_statuses.update_all(status: Katello::PurposeStatus::MATCHED) # 0 => 2
```

Reply:
@jturel updated with a migration. Tested that it works by manually editing my `PurposeStatus` records from 3 to 0 then running the migration and now they are a 2.

Reply:
The migration didn't work as expected for me. I ended up seeing ALL the purpose status objects have '2' as the status.

before migration:
```
 #<Katello::PurposeStatus:0x0000000013cc2410 id: 38, type: "Katello::PurposeStatus", status: 0, host_id: 1, reported_at: [FILTERED]>,
 #<Katello::PurposeSlaStatus:0x0000000013cc2000 id: 39, type: "Katello::PurposeSlaStatus", status: 2, host_id: 1, reported_at: [FILTERED]>]
```

after migration:
```
  #<Katello::PurposeStatus:0x0000000013b0ca08 id: 38, type: "Katello::PurposeStatus", status: 2, host_id: 1, reported_at: [FILTERED]>,
 #<Katello::PurposeSlaStatus:0x0000000013b0c800 id: 39, type: "Katello::PurposeSlaStatus", status: 2, host_id: 1, reported_at: [FILTERED]>]
```

suggested change to migration:
```
diff --git a/db/migrate/20200818192230_update_system_purpose_status.rb b/db/migrate/20200818192230_update_system_purpose_status.rb
index 32526b5..76cc7a0 100644
--- a/db/migrate/20200818192230_update_system_purpose_status.rb
+++ b/db/migrate/20200818192230_update_system_purpose_status.rb
@@ -3,10 +3,10 @@ class UpdateSystemPurposeStatus < ActiveRecord::Migration[6.0]
     purpose_types = Katello::HostStatusManager::PURPOSE_STATUS.map(&:to_s)
 
     # load both sets of host statuses and *then* update them to make sure we update the correct statuses
-    unknown_statuses = ::HostStatus::Status.where(type: purpose_types, status: 2)
-    matched_statuses = ::HostStatus::Status.where(type: purpose_types, status: 0)
+    unknown_statuses = ::HostStatus::Status.where(type: purpose_types, status: 2).pluck(:id)
+    matched_statuses = ::HostStatus::Status.where(type: purpose_types, status: 0).pluck(:id)
 
-    unknown_statuses.update_all(status: Katello::PurposeStatus::UNKNOWN) # 2 => 0
-    matched_statuses.update_all(status: Katello::PurposeStatus::MATCHED) # 0 => 2
+    ::HostStatus::Status.where(id: unknown_statuses).update_all(status: Katello::PurposeStatus::UNKNOWN) # 2 => 0
+    ::HostStatus::Status.where(id: matched_statuses).update_all(status: Katello::PurposeStatus::MATCHED) # 0 => 2
   end
 end

```

Reply:
@jturel updated, here is before and after:

Before:
```shell
"id"	"type"	"status"	"host_id"	"reported_at"
"2"	"Katello::PurposeSlaStatus"	"3"	"1"	"17/8/2020 18:09:50.072425"
"3"	"Katello::PurposeRoleStatus"	"3"	"1"	"17/8/2020 18:09:50.092667"
"4"	"Katello::PurposeUsageStatus"	"3"	"1"	"17/8/2020 18:09:50.115356"
"5"	"Katello::PurposeAddonsStatus"	"3"	"1"	"17/8/2020 18:09:50.183421"
"1"	"Katello::SubscriptionStatus"	"0"	"1"	"17/8/2020 18:09:53.219876"
"7"	"Katello::ErrataStatus"	"1"	"1"	"17/8/2020 18:09:59.541609"
"9"	"Katello::PurposeSlaStatus"	"3"	"2"	"17/8/2020 18:11:04.315267"
"10"	"Katello::PurposeRoleStatus"	"3"	"2"	"17/8/2020 18:11:04.336124"
"11"	"Katello::PurposeUsageStatus"	"3"	"2"	"17/8/2020 18:11:04.362418"
"12"	"Katello::PurposeAddonsStatus"	"3"	"2"	"17/8/2020 18:11:04.381024"
"8"	"Katello::SubscriptionStatus"	"2"	"2"	"17/8/2020 18:11:06.463509"
"14"	"Katello::ErrataStatus"	"1"	"2"	"17/8/2020 18:11:13.00389"
"16"	"Katello::PurposeSlaStatus"	"3"	"3"	"17/8/2020 18:12:02.345523"
"17"	"Katello::PurposeRoleStatus"	"3"	"3"	"17/8/2020 18:12:02.350816"
"18"	"Katello::PurposeUsageStatus"	"3"	"3"	"17/8/2020 18:12:02.35574"
"19"	"Katello::PurposeAddonsStatus"	"3"	"3"	"17/8/2020 18:12:02.361128"
"15"	"Katello::SubscriptionStatus"	"2"	"3"	"17/8/2020 18:12:04.964302"
"21"	"Katello::ErrataStatus"	"1"	"3"	"17/8/2020 18:12:10.045735"
"6"	"Katello::PurposeStatus"	"0"	"1"	"17/8/2020 18:09:50.200822"
"20"	"Katello::PurposeStatus"	"0"	"3"	"17/8/2020 18:12:02.366202"
"13"	"Katello::PurposeStatus"	"0"	"2"	"17/8/2020 18:11:04.397999"
```

After updated migration:
```shell
"id"	"type"	"status"	"host_id"	"reported_at"
"2"	"Katello::PurposeSlaStatus"	"3"	"1"	"17/8/2020 18:09:50.072425"
"3"	"Katello::PurposeRoleStatus"	"3"	"1"	"17/8/2020 18:09:50.092667"
"4"	"Katello::PurposeUsageStatus"	"3"	"1"	"17/8/2020 18:09:50.115356"
"5"	"Katello::PurposeAddonsStatus"	"3"	"1"	"17/8/2020 18:09:50.183421"
"1"	"Katello::SubscriptionStatus"	"0"	"1"	"17/8/2020 18:09:53.219876"
"7"	"Katello::ErrataStatus"	"1"	"1"	"17/8/2020 18:09:59.541609"
"9"	"Katello::PurposeSlaStatus"	"3"	"2"	"17/8/2020 18:11:04.315267"
"10"	"Katello::PurposeRoleStatus"	"3"	"2"	"17/8/2020 18:11:04.336124"
"11"	"Katello::PurposeUsageStatus"	"3"	"2"	"17/8/2020 18:11:04.362418"
"12"	"Katello::PurposeAddonsStatus"	"3"	"2"	"17/8/2020 18:11:04.381024"
"8"	"Katello::SubscriptionStatus"	"2"	"2"	"17/8/2020 18:11:06.463509"
"14"	"Katello::ErrataStatus"	"1"	"2"	"17/8/2020 18:11:13.00389"
"16"	"Katello::PurposeSlaStatus"	"3"	"3"	"17/8/2020 18:12:02.345523"
"17"	"Katello::PurposeRoleStatus"	"3"	"3"	"17/8/2020 18:12:02.350816"
"18"	"Katello::PurposeUsageStatus"	"3"	"3"	"17/8/2020 18:12:02.35574"
"19"	"Katello::PurposeAddonsStatus"	"3"	"3"	"17/8/2020 18:12:02.361128"
"15"	"Katello::SubscriptionStatus"	"2"	"3"	"17/8/2020 18:12:04.964302"
"21"	"Katello::ErrataStatus"	"1"	"3"	"17/8/2020 18:12:10.045735"
"6"	"Katello::PurposeStatus"	"2"	"1"	"17/8/2020 18:09:50.200822"
"20"	"Katello::PurposeStatus"	"2"	"3"	"17/8/2020 18:12:02.366202"
"13"	"Katello::PurposeStatus"	"2"	"2"	"17/8/2020 18:11:04.397999"
```

Reply:
Your Before state didn't have any of the purpose statuses with status == 2, so we can't be sure that this worked fully. Can you check with some purpose states in 0 and 2, and make sure the status is swapped between them after migration?

Reply:
Updated testing results:

Before:

```shell
"id"	"type"	"status"	"host_id"	"reported_at"
"1"	"Katello::SubscriptionStatus"	"0"	"1"	"17/8/2020 18:09:53.219876"
"7"	"Katello::ErrataStatus"	"1"	"1"	"17/8/2020 18:09:59.541609"
"8"	"Katello::SubscriptionStatus"	"2"	"2"	"17/8/2020 18:11:06.463509"
"14"	"Katello::ErrataStatus"	"1"	"2"	"17/8/2020 18:11:13.00389"
"15"	"Katello::SubscriptionStatus"	"2"	"3"	"17/8/2020 18:12:04.964302"
"21"	"Katello::ErrataStatus"	"1"	"3"	"17/8/2020 18:12:10.045735"
"2"	"Katello::PurposeSlaStatus"	"2"	"1"	"17/8/2020 18:09:50.072425"
"3"	"Katello::PurposeRoleStatus"	"2"	"1"	"17/8/2020 18:09:50.092667"
"4"	"Katello::PurposeUsageStatus"	"2"	"1"	"17/8/2020 18:09:50.115356"
"5"	"Katello::PurposeAddonsStatus"	"2"	"1"	"17/8/2020 18:09:50.183421"
"6"	"Katello::PurposeStatus"	"2"	"1"	"17/8/2020 18:09:50.200822"
"9"	"Katello::PurposeSlaStatus"	"0"	"2"	"17/8/2020 18:11:04.315267"
"10"	"Katello::PurposeRoleStatus"	"0"	"2"	"17/8/2020 18:11:04.336124"
"11"	"Katello::PurposeUsageStatus"	"0"	"2"	"17/8/2020 18:11:04.362418"
"12"	"Katello::PurposeAddonsStatus"	"0"	"2"	"17/8/2020 18:11:04.381024"
"13"	"Katello::PurposeStatus"	"0"	"2"	"17/8/2020 18:11:04.397999"
"16"	"Katello::PurposeSlaStatus"	"0"	"3"	"17/8/2020 18:12:02.345523"
"17"	"Katello::PurposeRoleStatus"	"0"	"3"	"17/8/2020 18:12:02.350816"
"18"	"Katello::PurposeUsageStatus"	"0"	"3"	"17/8/2020 18:12:02.35574"
"19"	"Katello::PurposeAddonsStatus"	"0"	"3"	"17/8/2020 18:12:02.361128"
"20"	"Katello::PurposeStatus"	"0"	"3"	"17/8/2020 18:12:02.366202"
```

After:

```shell
"id"	"type"	"status"	"host_id"	"reported_at"
"1"	"Katello::SubscriptionStatus"	"0"	"1"	"17/8/2020 18:09:53.219876"
"7"	"Katello::ErrataStatus"	"1"	"1"	"17/8/2020 18:09:59.541609"
"8"	"Katello::SubscriptionStatus"	"2"	"2"	"17/8/2020 18:11:06.463509"
"14"	"Katello::ErrataStatus"	"1"	"2"	"17/8/2020 18:11:13.00389"
"15"	"Katello::SubscriptionStatus"	"2"	"3"	"17/8/2020 18:12:04.964302"
"21"	"Katello::ErrataStatus"	"1"	"3"	"17/8/2020 18:12:10.045735"
"2"	"Katello::PurposeSlaStatus"	"0"	"1"	"17/8/2020 18:09:50.072425"
"3"	"Katello::PurposeRoleStatus"	"0"	"1"	"17/8/2020 18:09:50.092667"
"4"	"Katello::PurposeUsageStatus"	"0"	"1"	"17/8/2020 18:09:50.115356"
"5"	"Katello::PurposeAddonsStatus"	"0"	"1"	"17/8/2020 18:09:50.183421"
"6"	"Katello::PurposeStatus"	"0"	"1"	"17/8/2020 18:09:50.200822"
"9"	"Katello::PurposeSlaStatus"	"2"	"2"	"17/8/2020 18:11:04.315267"
"10"	"Katello::PurposeRoleStatus"	"2"	"2"	"17/8/2020 18:11:04.336124"
"11"	"Katello::PurposeUsageStatus"	"2"	"2"	"17/8/2020 18:11:04.362418"
"12"	"Katello::PurposeAddonsStatus"	"2"	"2"	"17/8/2020 18:11:04.381024"
"13"	"Katello::PurposeStatus"	"2"	"2"	"17/8/2020 18:11:04.397999"
"16"	"Katello::PurposeSlaStatus"	"2"	"3"	"17/8/2020 18:12:02.345523"
"17"	"Katello::PurposeRoleStatus"	"2"	"3"	"17/8/2020 18:12:02.350816"
"18"	"Katello::PurposeUsageStatus"	"2"	"3"	"17/8/2020 18:12:02.35574"
"19"	"Katello::PurposeAddonsStatus"	"2"	"3"	"17/8/2020 18:12:02.361128"
"20"	"Katello::PurposeStatus"	"2"	"3"	"17/8/2020 18:12:02.366202"
```